service/history/nDCTransactionMgr.go

```diff
@@ -232,83 +244,119 @@ func (r *nDCTransactionMgrImpl) backfillWorkflow(
 		return err
 	}
 
-	mode := persistence.UpdateWorkflowModeUpdateCurrent
-	transactionPolicy := transactionPolicyPassive
-	// since we are not rebuilding the mutable state then we
-	// can trust the result from IsCurrentWorkflowGuaranteed
-	if !targetWorkflow.getMutableState().IsCurrentWorkflowGuaranteed() {
-		executionInfo := targetWorkflow.getMutableState().GetExecutionInfo()
-		domainID := executionInfo.DomainID
-		workflowID := executionInfo.WorkflowID
-		runID := executionInfo.RunID
-
-		currentRunID, err := r.getCurrentWorkflowRunID(
-			ctx,
-			domainID,
-			workflowID,
-		)
-		if err != nil {
-			return err
-		}
-		if currentRunID != runID {
-			mode = persistence.UpdateWorkflowModeBypassCurrent
-		}
+	updateMode, transactionPolicy, err := r.backfillWorkflowEventsReapply(
+		ctx,
+		targetWorkflow,
+		targetWorkflowEvents,
+	)
+	if err != nil {
+		return err
 	}
 
+	return targetWorkflow.getContext().updateWorkflowExecutionWithNew(
+		now,
+		updateMode,
+		nil,
+		nil,
+		transactionPolicy,
+		nil,
+	)
+}
+
+func (r *nDCTransactionMgrImpl) backfillWorkflowEventsReapply(
+	ctx ctx.Context,
+	targetWorkflow nDCWorkflow,
+	targetWorkflowEvents *persistence.WorkflowEvents,
+) (persistence.UpdateWorkflowMode, transactionPolicy, error) {
+
+	isCurrentWorkflow, err := r.isWorkflowCurrent(ctx, targetWorkflow)
+	if err != nil {
+		return 0, transactionPolicyActive, err
+	}
+	isWorkflowRunning := targetWorkflow.getMutableState().IsWorkflowExecutionRunning()
 	targetWorkflowActiveCluster := r.clusterMetadata.ClusterNameForFailoverVersion(
-		targetWorkflow.getMutableState().GetCurrentVersion(),
+		targetWorkflow.getMutableState().GetDomainEntry().GetFailoverVersion(),
 	)
 	currentCluster := r.clusterMetadata.GetCurrentClusterName()
+	isActiveCluster := targetWorkflowActiveCluster == currentCluster
 
-	// workflow events reapplication to self (self workflow being current workflow)
-	// we need to handle 2 cases
-	// 1. workflow still running -> just reapply
-	// 2. workflow closed -> TODO wait until https://github.com/uber/cadence/issues/2420
-	//  NOTE: also remember that workflow reset will acquire a lock on current workflow (this)
-
-	// simple case workflow still running (implies current workflow),
-	// no workflow reset necessary
-	if targetWorkflowActiveCluster == currentCluster {
-		// target workflow is active && target workflow is current workflow
```

Comment:
I think this comment is useful, unless it is no longer relevant, please don't delete this.